tests/test_common.py

```diff
@@ -0,0 +1,276 @@
+
+from app.common import (
+    OTAFileCacheControl,
+    SimpleTasksTracker,
+    copytree_identical,
+    file_sha256,
+    re_symlink_atomic,
+    read_from_file,
+    subprocess_call,
+    subprocess_check_output,
+    verify_file,
+    write_to_file_sync,
+)
+from tests.utils import compare_dir
+
+logger = logging.getLogger(__name__)
+
+_TEST_FILE_CONTENT = "123456789abcdefgh" * 3000
+_TEST_FILE_SHA256 = sha256(_TEST_FILE_CONTENT.encode()).hexdigest()
+_TEST_FILE_LENGTH = len(_TEST_FILE_CONTENT)
+
+
+@pytest.fixture
+def file_t(tmp_path: Path) -> Tuple[str, str, int]:
+    """A fixture that returns a path to a test file and its sha256."""
+    test_f = tmp_path / "test_file"
+    test_f.write_text(_TEST_FILE_CONTENT)
+    return str(test_f), _TEST_FILE_SHA256, _TEST_FILE_LENGTH
+
+
+@pytest.mark.parametrize(
+    "_input, expected",
+    (
+        (
+            "use_cache,retry_caching",
+            {OTAFileCacheControl.use_cache, OTAFileCacheControl.retry_caching},
+        ),
+        ("no_cache", {OTAFileCacheControl.no_cache}),
+    ),
+)
+def test_parse_to_enum_set(_input: str, expected: Set[OTAFileCacheControl]):
+    assert OTAFileCacheControl.parse_to_enum_set(_input) == expected
+
+
+def test_file_sha256(file_t: Tuple[str, str, int]):
+    _path, _sha256, _ = file_t
+    assert file_sha256(_path) == _sha256
+
+
+def test_verify_file(tmp_path: Path, file_t: Tuple[str, str, int]):
+    _path, _sha256, _size = file_t
+    assert verify_file(Path(_path), _sha256, _size)
+    assert verify_file(Path(_path), _sha256, None)
+    assert not verify_file(Path(_path), _sha256, 123)
+    assert not verify_file(Path(_path), sha256().hexdigest(), 123)
+
+    # test over symlink file, verify_file should return False on symlink
+    _symlink = tmp_path / "test_file_symlink"
+    _symlink.symlink_to(_path)
+    assert not verify_file(_symlink, _sha256, None)
+
+
+def test_read_from_file(file_t: Tuple[str, str, int]):
+    _path, _, _ = file_t
+    # append some empty lines to the test_file
+    _append = "    \n      \n"
+    with open(_path, "a") as f:
+        f.write(_append)
+
+    assert read_from_file(_path) == _TEST_FILE_CONTENT
+    assert read_from_file("/non-existed", missing_ok=True, default="") == ""
+    assert read_from_file("/non-existed", missing_ok=True, default="abc") == "abc"
+    with pytest.raises(FileNotFoundError):
+        read_from_file("/non-existed", missing_ok=False)
+
+
+def test_write_to_file_sync(tmp_path: Path):
+    _path = tmp_path / "write_to_file"
+    write_to_file_sync(_path, _TEST_FILE_CONTENT)
+    assert _path.read_text() == _TEST_FILE_CONTENT
+
+
+def test_subprocess_call():
+    with pytest.raises(subprocess.CalledProcessError) as e:
+        subprocess_call("ls /non-existed", raise_exception=True)
+    _origin_e = e.value
+    assert _origin_e.returncode == 2
+    assert (
+        _origin_e.stderr.decode().strip()
+        == "ls: cannot access '/non-existed': No such file or directory"
+    )
```

Comment:
Please check https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-raises for detailed usage of `pytest.raises`.

Following is the example usage from the document:
```python
>>> with pytest.raises(ValueError) as exc_info:
...     raise ValueError("value must be 42")
>>> assert exc_info.type is ValueError
>>> assert exc_info.value.args[0] == "value must be 42"
```
The `with` block captures the exception info, and then we should examine the exception info **out of** the `with` block. 

Any lines after the line that raises the exception will not be executed, another example from document as follow:
```python
>>> value = 15
>>> with pytest.raises(ValueError) as exc_info:
...     if value > 10:
...         raise ValueError("value must be <= 10")
...     assert exc_info.type is ValueError  # this will not execute
```